### PR TITLE
Add photo to reward

### PR DESCRIPTION
**Pull request by K4sku**
Sprint 7 / task 1
Admin can add image to reward.

PR introduces active storage with google gcs bucket as production storage.
Admin can add single image to reward via new/edit form (partial), file type is limited to PNGs and JP(E)Gs. 
Images are displayed on admin's and employee's views. 
Photo image processing is handled by vips.

Specs:
- Reward model spec (file format validations)
- Adding image to reward system spec
- Displaying image by employee

Gems added: 
- google-cloud-storage
- image_processing

System dependencies:
- libvips

Code configurations:
- Active Storage model and migration, setup production service
- Schema update
- Encrypted GCS credentials
- Vips as variant processor

Changes outside of source code:
- GCS: create bucket, configure service account, generate json credentials
- Heroku: add apt and libvips buildpack annd libvips dependecies to Apt file 
- Github actions: add libvips to test workflow

reward photo

### CODE REVIEW

File: app/models/reward.rb [hunk 1]
```diff
@@ -1,10 +1,21 @@
...
   validates :categories, presence: true
+  validate :image_format
 
   paginates_per 3
+
+  private
+
+  def image_format
+    return unless image.attached?
+    return if %w[image/png image/jpg image/jpeg].include?(image.blob.content_type)
```
Comment (SzymonNadbrzezny) [inline app/models/reward.rb:17]: image/jpeg shouldn't be supported here

File: app/models/reward.rb [hunk 2]
```diff
@@ -1,10 +1,21 @@
 class Reward < ApplicationRecord
   has_many :category_rewards, dependent: :destroy
   has_many :categories, through: :category_rewards
+  has_one_attached :image
 
   validates :title, :description, :price, presence: true
   validates :price, numericality: { only_integer: true, greater_than_or_equal_to: 1 }
   validates :categories, presence: true
+  validate :image_format
```
Comment (SzymonNadbrzezny) [inline app/models/reward.rb:9]: It's great that you implemented it yourself. You could also use gem like `active_storage_validations`, but that's just a nitpick that is worth checking.

File: app/views/admins/rewards/_form.html.erb [hunk 3]
```diff
@@ -3,6 +3,24 @@
...
   <%= form.number_field :price, min: 1, step: 1 %>
+  
+
+  <div class="field">
+    <%= form.label :reward_image, "Image", class: 'label' %>
+    <div>
+      <%= image_tag @reward.image.variant(resize_to_limit: [200, nil]) if @reward.image.attached? && @reward.image.persisted? %>
+    </div>
+    <div class="control">
+      <%= form.file_field :image, accept: 'image/png,image/jpg,image/jpeg', test_id: "image_input" %>
```
Comment (SzymonNadbrzezny) [inline app/views/admins/rewards/_form.html.erb:14]: image/jpeg shouldn't be accepted here

File: app/views/admins/rewards/_form.html.erb [hunk 4]
```diff
@@ -3,6 +3,24 @@
...
+      <% if reward.errors.group_by_attribute.include?(:image) %>
```
Comment (SzymonNadbrzezny) [inline app/views/admins/rewards/_form.html.erb:15]: Great way to show errors that apply to images!

**Review by SzymonNadbrzezny — COMMENTED** [review]
Great job :+1: , left some suggestions

File: app/models/reward.rb (hunk 1 above)
Comment (K4sku) [reply app/models/reward.rb:17]: I strongly disagree with this, technically there is no file type jpg. Jpg is file extension for jpeg, it's not a separate file format. There is not MIME type of image/jpg.

https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Image_types#jpeg_joint_photographic_experts_group_image

File: app/views/admins/rewards/_form.html.erb (hunk 4 above)
Comment (K4sku) [reply app/views/admins/rewards/_form.html.erb:15]: Thanks for CR!

File: app/models/reward.rb (hunk 1 above)
Comment (SzymonNadbrzezny) [reply app/models/reward.rb:17]: I understand, but I believe that AC shows what extensions files can have. It doesn't talk about file type or MIME type.

Also thanks for that article, I actually didn't know that .jpg also stands for JPEG file type.

Comment (K4sku) [reply app/models/reward.rb:17]: On second thought, that makes one of types redundant on list.

Comment (ciejj) [inline app/models/reward.rb:17]: 👍

**Review by ciejj — APPROVED** [review]
Really nice PR!

**Pull request merged**